File: crates/rbac/tests-js/rbac.test.js
import { test, describe } from 'node:test';
import assert from 'node:assert/strict';
import { JsRbac, JsRbacBuilder, createRbacFromJson, checkPermission } from '../pkg/walrs_rbac.js';

describe('JsRbacBuilder', () => {
  describe('Constructor and Basic Building', () => {
    test('should create an empty RBAC builder', () => {
      const builder = new JsRbacBuilder();
      assert.ok(builder, 'Builder should be created');

      const rbac = builder.build();
      assert.ok(rbac instanceof JsRbac, 'Should build an RBAC instance');
    });

    test('should add a role with permissions', () => {
      const rbac = new JsRbacBuilder()
        .addRole('admin', ['manage.users'], null)
        .build();

      assert.ok(rbac.hasRole('admin'), 'RBAC should have the admin role');
      assert.ok(rbac.isGranted('admin', 'manage.users'), 'Admin should have manage.users');
    });

    test('should add a role with children', () => {
      const rbac = new JsRbacBuilder()
        .addRole('guest', ['read'], null)
        .addRole('user', ['write'], ['guest'])
        .build();

      assert.ok(rbac.hasRole('guest'), 'RBAC should have guest role');
      assert.ok(rbac.hasRole('user'), 'RBAC should have user role');
      assert.ok(rbac.isGranted('user', 'write'), 'User should have write permission');
      assert.ok(rbac.isGranted('user', 'read'), 'User should inherit read from guest');
      assert.ok(!rbac.isGranted('guest', 'write'), 'Guest should not have write');
    });
  });

  describe('Permission Inheritance', () => {
    test('should support deep inheritance', () => {
      const rbac = new JsRbacBuilder()
        .addRole('guest', ['read.public'], null)
        .addRole('user', ['write.post'], ['guest'])
        .addRole('editor', ['edit.post'], ['user'])
        .addRole('admin', ['admin.panel'], ['editor'])
        .build();

      // Admin inherits all
      assert.ok(rbac.isGranted('admin', 'admin.panel'));
      assert.ok(rbac.isGranted('admin', 'edit.post'));
      assert.ok(rbac.isGranted('admin', 'write.post'));
      assert.ok(rbac.isGranted('admin', 'read.public'));

      // Editor doesn't have admin
      assert.ok(!rbac.isGranted('editor', 'admin.panel'));
      assert.ok(rbac.isGranted('editor', 'edit.post'));
      assert.ok(rbac.isGranted('editor', 'write.post'));

      // Guest only has read
      assert.ok(rbac.isGranted('guest', 'read.public'));
      assert.ok(!rbac.isGranted('guest', 'write.post'));
    });

    test('should support multiple children', () => {
      const rbac = new JsRbacBuilder()
        .addRole('reader', ['read'], null)
        .addRole('writer', ['write'], null)
        .addRole('admin', ['admin'], ['reader', 'writer'])
        .build();

      assert.ok(rbac.isGranted('admin', 'read'));
      assert.ok(rbac.isGranted('admin', 'write'));
      assert.ok(rbac.isGranted('admin', 'admin'));
    });
  });

  describe('addPermission', () => {
    test('should add permission to existing role', () => {
      const rbac = new JsRbacBuilder()
        .addRole('user', ['read'], null)
        .addPermission('user', 'write')
        .build();

      assert.ok(rbac.isGranted('user', 'read'));
      assert.ok(rbac.isGranted('user', 'write'));
    });
  });

  describe('addChild', () => {
    test('should add child relationship', () => {
      const rbac = new JsRbacBuilder()
        .addRole('editor', ['edit'], null)
        .addRole('admin', ['admin'], null)
        .addChild('admin', 'editor')
        .build();

      assert.ok(rbac.isGranted('admin', 'edit'));
      assert.ok(rbac.isGranted('admin', 'admin'));
    });
  });

  describe('fromJson', () => {
    test('should create builder from JSON', () => {
      const json = JSON.stringify({
        roles: [
          ['guest', ['read'], null],
          ['admin', ['manage'], ['guest']]
        ]
      });

      const builder = JsRbacBuilder.fromJson(json);
      const rbac = builder.build();

      assert.ok(rbac.isGranted('admin', 'manage'));
      assert.ok(rbac.isGranted('admin', 'read'));
    });
  });

  describe('Error Handling', () => {
    test('should throw on invalid JSON', () => {
      assert.throws(() => {
        JsRbacBuilder.fromJson('invalid json');
      });
    });

    test('should throw on missing child role', () => {
      assert.throws(() => {
        new JsRbacBuilder()
          .addRole('admin', ['manage'], ['nonexistent'])
          .build();
      });
    });

    test('should throw on cycle', () => {
      assert.throws(() => {
        new JsRbacBuilder()
          .addRole('a', [], ['b'])
          .addRole('b', [], ['a'])
          .build();
      });
    });
  });
});

describe('JsRbac', () => {
  describe('Constructor', () => {
    test('should create empty RBAC', () => {
      const rbac = new JsRbac();
      assert.equal(rbac.roleCount(), 0);
    });
  });

  describe('fromJson', () => {
    test('should create RBAC from JSON', () => {
      const json = JSON.stringify({
        roles: [
          ['guest', ['read'], null],
          ['admin', ['manage'], ['guest']]
        ]
      });

      const rbac = JsRbac.fromJson(json);

      assert.ok(rbac.hasRole('guest'));
      assert.ok(rbac.hasRole('admin'));
      assert.ok(rbac.isGranted('admin', 'manage'));
      assert.ok(rbac.isGranted('admin', 'read'));
    });

    test('should throw on invalid JSON', () => {
      assert.throws(() => {
        JsRbac.fromJson('not json');
      });
    });
  });

  describe('isGranted', () => {
    test('should return false for non-existent role', () => {
      const rbac = new JsRbac();
      assert.equal(rbac.isGranted('nonexistent', 'anything'), false);
    });
  });

  describe('hasRole', () => {
    test('should return false for non-existent role', () => {
      const rbac = new JsRbac();
      assert.equal(rbac.hasRole('nonexistent'), false);
    });
  });
});

describe('Convenience Functions', () => {
  test('createRbacFromJson should create RBAC', () => {
    const json = JSON.stringify({
      roles: [
        ['admin', ['manage'], null]
      ]
    });

    const rbac = createRbacFromJson(json);
    assert.ok(rbac.isGranted('admin', 'manage'));
  });

  test('checkPermission should check permission directly', () => {
    const json = JSON.stringify({
      roles: [
        ['admin', ['manage'], null]
      ]
    });

    assert.equal(checkPermission(json, 'admin', 'manage'), true);
    assert.equal(checkPermission(json, 'admin', 'other'), false);
  });
});

File: crates/rbac/tests/rbac_test.rs
use walrs_rbac::{RbacBuilder, RbacError};

#[test]
fn test_basic_rbac() -> Result<(), RbacError> {
  let rbac = RbacBuilder::new()
    .add_role("guest", &["read.public"], None)?
    .add_role("user", &["write.post", "comment.post"], Some(&["guest"]))?
    .add_role("editor", &["edit.post", "publish.post"], Some(&["user"]))?
    .add_role("admin", &["admin.panel", "manage.users"], Some(&["editor"]))?
    .build()?;

  // Guest
  assert!(rbac.is_granted("guest", "read.public"));
  assert!(!rbac.is_granted("guest", "write.post"));
  assert!(!rbac.is_granted("guest", "admin.panel"));

  // User (inherits from guest)
  assert!(rbac.is_granted("user", "read.public"));
  assert!(rbac.is_granted("user", "write.post"));
  assert!(rbac.is_granted("user", "comment.post"));
  assert!(!rbac.is_granted("user", "edit.post"));
  assert!(!rbac.is_granted("user", "admin.panel"));

  // Editor (inherits from user)
  assert!(rbac.is_granted("editor", "read.public"));
  assert!(rbac.is_granted("editor", "write.post"));
  assert!(rbac.is_granted("editor", "edit.post"));
  assert!(rbac.is_granted("editor", "publish.post"));
  assert!(!rbac.is_granted("editor", "admin.panel"));

  // Admin (inherits from editor)
  assert!(rbac.is_granted("admin", "read.public"));
  assert!(rbac.is_granted("admin", "write.post"));
  assert!(rbac.is_granted("admin", "edit.post"));
  assert!(rbac.is_granted("admin", "publish.post"));
  assert!(rbac.is_granted("admin", "admin.panel"));
  assert!(rbac.is_granted("admin", "manage.users"));

  Ok(())
}

#[test]
fn test_multiple_children() -> Result<(), RbacError> {
  let rbac = RbacBuilder::new()
    .add_role("reader", &["read"], None)?
    .add_role("writer", &["write"], None)?
    .add_role("deleter", &["delete"], None)?
    .add_role(
      "superuser",
      &["manage"],
      Some(&["reader", "writer", "deleter"]),
    )?
    .build()?;

  assert!(rbac.is_granted("superuser", "read"));
  assert!(rbac.is_granted("superuser", "write"));
  assert!(rbac.is_granted("superuser", "delete"));
  assert!(rbac.is_granted("superuser", "manage"));

  assert!(!rbac.is_granted("reader", "write"));
  assert!(!rbac.is_granted("writer", "read"));

  Ok(())
}

#[test]
fn test_diamond_inheritance() -> Result<(), RbacError> {
  // base <- left, base <- right, left + right <- top
  let rbac = RbacBuilder::new()
    .add_role("base", &["base.perm"], None)?
    .add_role("left", &["left.perm"], Some(&["base"]))?
    .add_role("right", &["right.perm"], Some(&["base"]))?
    .add_role("top", &["top.perm"], Some(&["left", "right"]))?
    .build()?;

  assert!(rbac.is_granted("top", "top.perm"));
  assert!(rbac.is_granted("top", "left.perm"));
  assert!(rbac.is_granted("top", "right.perm"));
  assert!(rbac.is_granted("top", "base.perm"));

  Ok(())
}

#[test]
fn test_no_false_positives() -> Result<(), RbacError> {
  let rbac = RbacBuilder::new()
    .add_role("user", &["read"], None)?
    .add_role("admin", &["manage"], None)?
    .build()?;

  // They share no inheritance
  assert!(!rbac.is_granted("user", "manage"));
  assert!(!rbac.is_granted("admin", "read"));
  assert!(!rbac.is_granted("user", "nonexistent"));
  assert!(!rbac.is_granted("nonexistent", "read"));

  Ok(())
}

#[test]
fn test_cycle_detection_error() {
  let result = RbacBuilder::new()
    .add_role("a", &[], Some(&["b"]))
    .unwrap()
    .add_role("b", &[], Some(&["a"]))
    .unwrap()
    .build();

  assert!(result.is_err());
  match result {
    Err(RbacError::CycleDetected(_)) => {}
    other => panic!("Expected CycleDetected, got {:?}", other),
  }
}

#[test]
fn test_missing_child_error() {
  let result = RbacBuilder::new()
    .add_role("admin", &["manage"], Some(&["nonexistent"]))
    .unwrap()
    .build();

  assert!(result.is_err());
  match result {
    Err(RbacError::InvalidConfiguration(msg)) => {
      assert!(msg.contains("nonexistent"));
    }
    other => panic!("Expected InvalidConfiguration, got {:?}", other),
  }
}

#[test]
fn test_empty_rbac() -> Result<(), RbacError> {
  let rbac = RbacBuilder::new().build()?;
  assert_eq!(rbac.role_count(), 0);
  assert!(!rbac.has_role("anything"));
  assert!(!rbac.is_granted("anything", "anything"));
  Ok(())
}
